### PR TITLE
Scan: Handle NotFound case for Datasources and Oracle-scripts page

### LINKED ISSUE #2382: Handle NotFound case for Datasources and Oraclescripts page when The ID doesn't exit
- When loading the data, we should check the data source or oracle script is the valid one.
- Use `NotFound` Page when it's no data.

### PR DESCRIPTION
**Pull request by Pitchanai**
Fixed: #2382 

## Implementation details
- Handle NotFound case for Datasources and Oracle-scripts page when The ID doesn't existed.

## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [x] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [x] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: scan/src/pages/DataSourceIndexPage.re [hunk 1]
```diff
@@ -16,91 +16,101 @@ module Styles = {
...
-            <Heading value="Description" size=Heading.H5 marginBottom=16 />
+
+  switch (dataSourceSub) {
+  | NoData => <NotFound />
+  | _ =>
+    <Section pbSm=0>
+      <div className=CssHelper.container>
+        <Row.Grid marginBottom=40 marginBottomSm=16>
+          <Col.Grid>
+            <Heading value="Data Source" size=Heading.H4 marginBottom=40 marginBottomSm=24 />
```
Comment (pzshine) [inline scan/src/pages/DataSourceIndexPage.re]: Can we move all of the elements from the 23rd line as a component to make it easier to maintain?
@evilpeach What do you think?

File: scan/src/pages/OracleScriptIndexPage.re [hunk 2]
```diff
@@ -21,196 +21,210 @@ module Styles = {
...
-                      CssHelper.flexBox(~justify=`center, ()),
-                      Styles.titleSpacing,
-                    ])}>
-                    <Heading value="Response time" size=Heading.H4 align=Heading.Center />
+  switch (oracleScriptSub) {
+  | NoData => <NotFound />
+  | _ =>
+    <Section pbSm=0>
+      <div className=CssHelper.container>
+        <Heading value="Oracle Script" size=Heading.H4 marginBottom=40 marginBottomSm=24 />
```
Comment (pzshine) [inline scan/src/pages/OracleScriptIndexPage.re]: Here as well
@evilpeach What do you think?

**Review by pzshine — CHANGES_REQUESTED** [review]

File: scan/src/pages/DataSourceIndexPage.re (hunk 1 above)
Comment (pzshine) [reply scan/src/pages/DataSourceIndexPage.re]: Can be made in the same file

Comment (Pitchanai) [reply scan/src/pages/DataSourceIndexPage.re]: Updated! Please check again. @pzshine

File: scan/src/pages/OracleScriptIndexPage.re (hunk 2 above)
Comment (Pitchanai) [reply scan/src/pages/OracleScriptIndexPage.re]: Here's too. @pzshine

**Review by pzshine — APPROVED** [review]
LGTM @evilpeach  please review again

File: CHANGELOG_UNRELEASED.md [hunk 3]
```diff
@@ -26,6 +26,7 @@
 
 ### Scan
 
+- (impv) [\#2695](https://github.com/bandprotocol/bandchain/pull/2714) Handle NotFound case for Datasources and Oracle-scripts page
```
Comment (evilpeach) [inline CHANGELOG_UNRELEASED.md]: change to 2714

File: scan/src/pages/DataSourceIndexPage.re [hunk 4]
```diff
@@ -13,94 +13,115 @@ module Styles = {
     style([borderBottom(`px(1), `solid, Colors.gray9), paddingBottom(`px(16))]);
 };
 
+module Content = {
+  [@react.component]
+  let make =
+      (
+        ~dataSourceSub: ApolloHooks.Subscription.variant(BandScan.DataSourceSub.t),
+        ~dataSourceID,
+        ~hashtag: Route.data_source_tab_t,
```
Comment (evilpeach) [inline scan/src/pages/DataSourceIndexPage.re]: It's no need to declare the type. I think the compiler can imply it.

File: scan/src/pages/OracleScriptIndexPage.re [hunk 5]
```diff
@@ -17,200 +17,225 @@ module Styles = {
...
-                      Styles.titleSpacing,
-                    ])}>
-                    <Heading value="Response time" size=Heading.H4 align=Heading.Center />
+module Content = {
+  [@react.component]
+  let make =
+      (
+        ~oracleScriptSub: ApolloHooks.Subscription.variant(BandScan.OracleScriptSub.t),
+        ~oracleScriptID,
+        ~hashtag: Route.oracle_script_tab_t,
```
Comment (evilpeach) [inline scan/src/pages/OracleScriptIndexPage.re]: Here as well

**Review by evilpeach — CHANGES_REQUESTED** [review]

**Review by evilpeach — CHANGES_REQUESTED** [review]
If the user inputs the invalid Data Source ID like `/data-source/12d`, it'll error. Let's change `int_of_string` to `int_of_string_opt` on `Route.re`

File: scan/src/utils/Route.re [hunk 6]
```diff
@@ -55,7 +55,11 @@ let fromUrl = (url: ReasonReactRouter.url) =>
       | "revisions" => DataSourceRevisions
       | _ => DataSourceRequests
     );
-    DataSourceIndexPage(dataSourceID |> int_of_string, urlHash(hash));
+    switch (dataSourceID |> int_of_string_opt) {
+    | None => NotFound
+    | Some(dataSourceIDInt) => DataSourceIndexPage(dataSourceIDInt, urlHash(hash))
```
Comment (pzshine) [inline scan/src/utils/Route.re]: Can you swap the order of some and none?

File: scan/src/utils/Route.re [hunk 7]
```diff
@@ -66,7 +70,11 @@ let fromUrl = (url: ReasonReactRouter.url) =>
       | "revisions" => OracleScriptRequests
       | _ => OracleScriptRequests
     );
-    OracleScriptIndexPage(oracleScriptID |> int_of_string, urlHash(hash));
+    switch (oracleScriptID |> int_of_string_opt) {
+    | None => NotFound
+    | Some(oracleScriptIDInt) => OracleScriptIndexPage(oracleScriptIDInt, urlHash(hash))
```
Comment (pzshine) [inline scan/src/utils/Route.re]: Can you swap the order of some and none?

**Review by pzshine — CHANGES_REQUESTED** [review]

**Review by evilpeach — APPROVED** [review]

**Review by pzshine — APPROVED** [review]

**Pull request merged**